### PR TITLE
add pass for linalg matmultransposeb op

### PR DESCRIPTION
**Pull request by zhxzh-2001**
add Linalg MatMultransposebpass do the vectorization to speed up llama

### CODE REVIEW

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+#include <mlir/Dialect/Func/IR/FuncOps.h>
+#include <mlir/Dialect/Linalg/Transforms/Transforms.h>
+#include <mlir/IR/Dialect.h>
+#include <mlir/IR/Operation.h>
+#include <mlir/IR/TypeUtilities.h>
+#include <mlir/IR/Value.h>
+#include <mlir/Pass/Pass.h>
+
+#include "Utils/Utils.h"
+#include <iostream>
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:31]: Is `iostream` library necessary here?

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+using namespace mlir;
+using namespace vector;
+using namespace std;
+//===----------------------------------------------------------------------===//
+// Rewrite Pattern
+//===----------------------------------------------------------------------===//
+
+namespace {
+class MatMul_TransposeB_VecPattern : public ConversionPattern{
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:40]: MatMul_TransposeB_VecPattern -> MatMulTransposeBVecPattern

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            rewriter.create<arith::ConstantIndexOp>(loc,vecsize);
+        
+        const Value aRow = rewriter.create<memref::DimOp>(loc,A,c0);
+        const Value bRow = rewriter.create<memref::DimOp>(loc,B,c0);
+        const Value bCol = rewriter.create<memref::DimOp>(loc,B,c1);
+        
+        SmallVector<Value,8> lowerBounds(2,c0);
+        SmallVector<Value,8> uperBounds {aRow,bRow/*bCol*/};
+        SmallVector<int64_t,8> steps{1,1};
+        //TODO
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:74]: Please provide more detailed information about the `TODO`.

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        //generating affine code
+        affine::buildAffineLoopNest(
+            rewriter,loc,lowerBounds,uperBounds,steps,
+            [&](OpBuilder &builder,Location loc,ValueRange ivs){
+                //Value sum_0 = builder.create<mlir::arith::ConstantOp>(
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:79]: If we do not need this line, please remove it.

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                            loc,mlir::vector::CombiningKind::ADD,resvec
+                        );
+                        Value sum = builder.create<arith::AddFOp>(loc,itrArgs[0],ans);
+                        builder.create<affine::AffineYieldOp>(loc,sum);
+                    }
+                );
+                builder.create<affine::AffineStoreOp>(loc,sum.getResult(0),C,ValueRange{ivs[0],ivs[1]});
+            }
+        );
+        // clang-format on
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:107]: Did not find the matched `clang-format off`.

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        rewriter.eraseOp(op);
+        return success();   
+    }
+private:
+    int64_t vecsize;
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:112]: vecsize -> vecSize

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+//===----------------------------------------------------------------------===//
+// MatMulVectorizationPass
+//===----------------------------------------------------------------------===//
+
+namespace{
+    class MatMul_TransposeB_VecPass
+        :public PassWrapper<MatMul_TransposeB_VecPass,OperationPass<ModuleOp>>{
+public:
+    MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(MatMul_TransposeB_VecPass)
+    StringRef getArgument() const final{ return "matul-vectorization2"; }
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:125]: `matul-vectorization2` is not a good name for the pass.

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    signalPassFailure();
+}
+
+namespace mlir {
+namespace buddy {
+void registerMatMul_TransposeB_VecPass() {
+  PassRegistration<MatMul_TransposeB_VecPass>();
+}
+} // namespace buddy
+} // namespace mlir
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:164]: Please add an empty line here.

**Review by zhanghb97 — COMMENTED** [review]
Giving quick feedback. Additionally, please add test and example for your vectorization pass.

File: examples/BuddyMatmul/linalg-transposematmulb-f32.mlir
Comment (zhanghb97) [inline examples/BuddyMatmul/linalg-transposematmulb-f32.mlir:1]: Please add a test to this file. See [here](https://github.com/buddy-compiler/buddy-mlir/blob/main/examples/BuddyMatmul/linalg-batchmatmul-f32.mlir) as an example.

File: examples/BuddyMatmul/makefile [hunk 9]
```diff
@@ -35,3 +35,22 @@ linalg-batchmatmul-f32-run:
 		-reconcile-unrealized-casts | \
 	${MLIR_CPU_RUNNER} ${OPT_FLAG} -e main -entry-point-result=void \
 		-shared-libs=${MLIR_RUNNER_UTILS} -shared-libs=${MLIR_C_RUNNER_UTILS}
+
+linalg-transposematmulb-f32-run:
```
Comment (zhanghb97) [inline examples/BuddyMatmul/makefile:39]: Split `transpose`, `matmul`, and `b`. (`transposematmulb` -> `matmul-transpose-b`)

File: examples/BuddyMatmul/makefile [hunk 10]
```diff
@@ -35,3 +35,22 @@ linalg-batchmatmul-f32-run:
...
+	@${BUDDY_OPT} ./linalg-transposematmulb-f32.mlir\
+		-transpose_matmul_bvectorization \
```
Comment (zhanghb97) [inline examples/BuddyMatmul/makefile:41]: `-transpose_matmul_bvectorization` -> `-matmul_transpose_b_vectorization`

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:1]: Please use `clang-format` to format the file.

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+// MatMulVectorizationPass
+//===----------------------------------------------------------------------===//
+
+namespace{
+    class MatMulTransposeBVecPass
+        :public PassWrapper<MatMulTransposeBVecPass,OperationPass<ModuleOp>>{
+public:
+    MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(MatMulTransposeBVecPass)
+    StringRef getArgument() const final{ return "transpose_matmul_bvectorization"; }
+    StringRef getDescription() const final { return "MatMul Vectorization second version.MatMul receive tensortype oprands."; }
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:123]: "MatMul Vectorization second version.MatMul receive tensortype oprands." This is not a good description for this pass.

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+//===----------------------------------------------------------------------===//
+// MatMulVectorizationPass
+//===----------------------------------------------------------------------===//
+
+namespace{
+    class MatMulTransposeBVecPass
+        :public PassWrapper<MatMulTransposeBVecPass,OperationPass<ModuleOp>>{
+public:
+    MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(MatMulTransposeBVecPass)
+    StringRef getArgument() const final{ return "transpose_matmul_bvectorization"; }
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:122]: `transpose_matmul_bvectorization` -> `matmul_transpose_b_vectorization`

**Review by zhanghb97 — COMMENTED** [review]
Put the Python/Graph part in a separate PR. This PR only adds matmul transpose vectorization.

**Comment by xlinsist** [comment]
I encountered this error when reviewing PR. Please take another look. After changing buddy-opt related things, you need to use `ninja && ninja check-buddy` in the build directory.

image

File: examples/BuddyMatmul/linalg-transposematmulb-f32.mlir [hunk 13]
```diff
@@ -0,0 +1,79 @@
+// RUN: buddy-opt %s \
+// RUN:     -matmul_transpose_b_vectorization \
```
Comment (xlinsist) [inline examples/BuddyMatmul/linalg-transposematmulb-f32.mlir:2]: Recommend to change `-matmul_transpose_b_vectorization` to `-matmul-transpose-b-vectorization`

**Comment by xlinsist** [comment]
I encountered this error after using `ninja && ninja check-buddy` in the build directory. Could you reproduce it in your machine?

image

File: examples/BuddyMatmul/linalg-transposematmulb-f32.mlir [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+  %printed_m5 = memref.cast %m5 : memref<?x?xf32> to memref<*xf32>
+
+  // CHECK: Unranked Memref base@ = {{.*}} rank = 2 offset = 0 sizes = [3, 3] strides = [3, 1] data =
+  // CHECK-NEXT: [
+  // CHECK: [3{{(, 3)*}}]
+  call @printMemrefF32(%printed_m5) : (memref<*xf32>) -> ()
+
+  return
+}
```
Comment (xlinsist) [inline examples/BuddyMatmul/linalg-transposematmulb-f32.mlir:75]: Add a new line at the end of the file.

File: tools/buddy-opt/buddy-opt.cpp [hunk 15]
```diff
@@ -80,6 +80,10 @@ void registerLowerSchePass();
 void registerFuncBufferizeDynamicOffsetPass();
 void registerConvertMemcpyToGPUPass();
 void registerLegalizeShmemOutliningPass();
+void registerMatMul_TransposeB_VecPass();
```
Comment (xlinsist) [inline tools/buddy-opt/buddy-opt.cpp:83]: What is the difference between `registerMatMul_TransposeB_VecPass` and `registerMatMulTransposeBVecPass`? Does this code have any particular meaning?

File: midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+    matchAndRewrite(Operation *op,ArrayRef<Value> /*operands*/,
+                ConversionPatternRewriter &rewriter) const override{
+        auto loc = op->getLoc();
+        auto ctx = op->getContext();
+        // Get input A, B, C.
+        Value A = op->getOperand(0);
+        Value B = op->getOperand(1);
+        Value C = op->getOperand(2);
+
+        // Get shape of input and output
```
Comment (xlinsist) [inline midend/lib/Conversion/MatMulOptimization/MatMulTransposeBVec.cpp:57]: typo: add period

File: frontend/Python/ops/linalg.py [hunk 17]
```diff
@@ -1968,6 +1986,7 @@ def gt_op(node: GtOp, symbol_table):
 
 ops_registry = {
     "MatmulOp": matmul_op,
+    "transpose_Matmul_fusedOp": transpose_matmul_fused_op,
```
Comment (xlinsist) [inline frontend/Python/ops/linalg.py:1989]: Recommend naming "TransposeMatmulFusedOp"

File: tools/buddy-opt/buddy-opt.cpp (hunk 15 above)
Comment (zhxzh-2001) [reply tools/buddy-opt/buddy-opt.cpp:83]: sorry,i typed an extra function.there is no pass named  "registerMatMul_TransposeB_VecPass"

Comment (zhxzh-2001) [reply tools/buddy-opt/buddy-opt.cpp:83]: I'll delete it

**Comment by zhxzh-2001** [comment]
yes, it could be verified in my next pr.This pass is for an op which will be used after the next modification of graph,and will help with llama.

**Pull request merged**